### ISSUE TITLE
GDEF glyph classes in IR

### ISSUE THREAD

**Issue by cmyr** [googlefonts/fontc#767]
In order to properly generate mark attachment rules, we need to know the GDEF class assignments for glyphs. This information can be specified in various ways:

- in glyphs.app, it is a property on the individual glyphs.
- in UFO, it is an optional field in the root lib, [`public.openTypeCategories`](https://unifiedfontobject.org/versions/ufo3/lib.plist/#publicopentypecategories)

In addition to these locations, the GDEF classes can also be specified via [FEA syntax](http://adobe-type-tools.github.io/afdko/OpenTypeFeatureFileSpecification.html#9b-gdef-table), and this is the main source of problems. Basically: we would _really_ like to have a single source of truth here. There is also a chicken-egg problem, where for _generated_ features we need to know these class rules.

So: I would like to decide on some standardized behaviour, and I would like to move these classdefs into IR. 

- for glyphs.app sources, the source of truth will always be per-glyph categories.
    - Q: _are these always available, or only if explicitly set by the user? Otherwise I assume they are computed from `GlyphData.xml`?_
- for UFO/designspace sources, the source of truth will be `public.openTypeCategories` dictionary, if present.

Issues to consider (and my current sense of how to address them)

- what if a UFO does not have the `public.openTypeCategories`, but does have explicit classes defined in FEA? _we use the categories from the FEA_.
- what if a UFO has both, and they don't match? _print a warning and use the lib version_.
- what if UFO doesn't have either? _good question_. _Probably I would synthesize one from `GlyphData.xml`? Is there a reason this is bad?_

the FEA spec says that if the GDEF classes are not stated explicitly then they should be generated based on what glyphs appear in mark/base/lig mark attachment rules, but since we are generating most of these lookups automatically, _this really doesn't work for us_. We could infer these based on the lookups we generate, but this runs into possible problems if there are non-base glyphs that include an anchor that is used to attach a component, but which are not expected to participate in mark-to-base? @anthrotype mentioned this as a possible situation, although I'm not sure how common it is.

TL;DR:

- let's add GDEF glyph class defs to IR
- let's get it "not from FEA" by default, and fall back to getting it from FEA
- let's use `GlyphData.xml` to synthesize it if necessary (always for glyph.app except for manual overrides; only for UFO if no lib key _and_ no GDEF block in FEA)

**Comment by anthrotype** [comment]
> for glyphs.app sources, the source of truth will always be per-glyph categories.
Q: are these always available, or only if explicitly set by the user? Otherwise I assume they are computed from GlyphData.xml?

the category/subCategory glyph property are not always available in a .glyphs source file, they are present only if the user has explicitly set those (via CMD+ALT+I glyph info panel); otherwise it is assumed that the embedded default GlyphData.xml is consulted (looked up primarily by the glyph name, alternatively by unicode codepoint associated with the glyph IIRC).
One can also provide a modified GlyphData.xml as input to fontmake --glyph-data (see #766), which would override (and can be a subset of) the default data file.

> what if a UFO has both, and they don't match? print a warning and use the lib version.

Currently in ufo2ft, the FEA GlyphClassDefs would always take precedence over the public.openTypeCategories, since it is assumed to be kind of more specific/lower level; no warning is issued, they are checked in order and if FEA is present, the other isn't consulted.

https://github.com/googlefonts/ufo2ft/blob/4d9aca9d0976efdde1913ab0481f27ef33d5b15d/Lib/ufo2ft/featureWriters/baseFeatureWriter.py#L356-L376

Originally only the FEA GlyphClassDefs were available to be defined and some old fonts may use that approach, however I'd expect most fonts to only define the lib public.openTypeCategories, not the other (let alone both).

> what if UFO doesn't have either? good question. Probably I would synthesize one from GlyphData.xml? Is there a reason this is bad?

no, that would not match fontmake and would surprise users of UFO+DS. The GlyphData.xml is very Glyphs.app specific and should only be used for .glyphs sources. The glyph names defined in there are not a standard.
If there are no categories defined, the mark writer should assume that any potential glyph that has attaching anchors is either a base, or a mark, or a ligature (if it has the _1, _2, etc. suffixed anchors).

> this runs into possible problems if there are non-base glyphs that include an anchor that is used to attach a component, but which are not expected to participate in mark-to-base?

potentially there is this risk, but it's not really an issue (worst case we generate more mark rules than one may need), but the font developer has a documented way to ensure that only the selected group of bases, marks and ligatures are considered for mark/mkmk feature generation (by definining a public.openTypeCategories in the lib).
Also, I don't think it's common from a _pure_ ufo+DS workflow (not converted from .glyphs) to have anchors that aren't meant to be used for mark/mkmk feature but only for automatic component alignment, since the latter is another specific Glyphs.app-only feature as far as I am aware.

> let's add GDEF glyph class defs to IR

SGTM

> let's get it "not from FEA" by default, and fall back to getting it from FEA

unfortunately, FEA currently would take precendence so we should match that.

> let's use GlyphData.xml to synthesize it if necessary (always for glyph.app except for manual overrides; only for UFO if no lib key and no GDEF block in FEA)

we should only use GlyphData.xml for .glyphs sources

**Comment by dfrg** [comment]
> There is a 'ligature' category, but no 'component' category, which makes sense I guess?

Right. A glyph becomes a component when it is consumed by a ligature substitution so it is purely contextual.

**Comment by cmyr** [comment]
> > for glyphs.app sources, the source of truth will always be per-glyph categories.
> > Q: are these always available, or only if explicitly set by the user? Otherwise I assume they are computed from GlyphData.xml?
> 
> the category/subCategory glyph property are not always available in a .glyphs source file, they are present only if the user has explicitly set those (via CMD+ALT+I glyph info panel); otherwise it is assumed that the embedded default GlyphData.xml is consulted (looked up primarily by the glyph name, alternatively by unicode codepoint associated with the glyph IIRC). One can also provide a modified GlyphData.xml as input to fontmake --glyph-data (see #766), which would override (and can be a subset of) the default data file.

Okay interesting, I hadn't seen that issue.

> > what if a UFO has both, and they don't match? print a warning and use the lib version.
> 
> Currently in ufo2ft, the FEA GlyphClassDefs would always take precedence over the public.openTypeCategories, since it is assumed to be kind of more specific/lower level; no warning is issued, they are checked in order and if FEA is present, the other isn't consulted.
> 
> https://github.com/googlefonts/ufo2ft/blob/4d9aca9d0976efdde1913ab0481f27ef33d5b15d/Lib/ufo2ft/featureWriters/baseFeatureWriter.py#L356-L376
> 
> Originally only the FEA GlyphClassDefs were available to be defined and some old fonts may use that approach, however I'd expect most fonts nowadays to only define the lib public.openTypeCategories, not the other (let alone both).

okay, good to know. It does seem to me that if they're both present it's probably a mistake of *some* kind? But this seems like a minor point.

> > what if UFO doesn't have either? good question. Probably I would synthesize one from GlyphData.xml? Is there a reason this is bad?
> 
> no, that would not match fontmake and would surprise users of UFO+DS. The GlyphData.xml is very Glyphs.app specific and should only be used for .glyphs sources. The glyph names defined in there are not a standard. If there are no categories defined, the mark writer should assume that any potential glyph that has attaching anchors is either a base, or a mark, or a ligature (if it has the _1, _2, etc. suffixed anchors).

👍 

> > let's add GDEF glyph class defs to IR
> 
> SGTM
> 
> > let's get it "not from FEA" by default, and fall back to getting it from FEA
> 
> unfortunately, FEA currently would take precendence so we should match that.

> > let's use GlyphData.xml to synthesize it if necessary (always for glyph.app except for manual overrides; only for UFO if no lib key and no GDEF block in FEA)
> 
> we should only use GlyphData.xml for .glyphs sources

Okay so then this will be the situation as I understand it:

- for ufo+ds sources, we will use whatever is defined in the FEA, falling back to the openTypeCategories lib key.
- for glyphs.app, there are never explicitly defined GDEF classes, as far as I can tell? There are the Category/subcategory fields on individual glyphs, and these influence... something? I imagine it influences the glyphs code that generates mark lookups, and then I _imagine_ that the GDEF classes are defined automatically by whatever is compiling the glyphs FEA code, based inferring the glyphs from what appears in various kinds of lookups? I could verify this experimentally but it would be nice to have this confirmed otherwise.

**Comment by cmyr** [comment]
interesting, and glyphsLib is just using the glyph category/subcategory to determine the openTypeCategories? Is it using anything else? Is it setting the 'components' group?

As an aside, there also seems to be some special logic in glyphs for inferring whether a glyph is a ligature based on the name; e.g. if I create a new glyph with a name like `v_e_r_y_u_n_c_o_m_m_o_n_l_i_g` glyphs assigns it to the 'ligature' subcategory (but not `very_uncommon_lig`, but it _does_ do this for both `a_Abreve` and `a_Zbreve`, which latter indicates it isn't checking that each component is a known glyph name.) Is this logic matched in glyphsLib somewhere?

**Comment by anthrotype** [comment]
> glyphsLib is just using the glyph category/subcategory to determine the openTypeCategories?

https://github.com/googlefonts/glyphsLib/blob/e2ebf5b517d59bec0c9437da3a748c58f2999911/Lib/glyphsLib/builder/features.py#L205

> setting the 'components' group?

those are unused/useless, ignore

> inferring whether a glyph is a ligature based on the name

I think here https://github.com/googlefonts/glyphsLib/blob/e2ebf5b517d59bec0c9437da3a748c58f2999911/Lib/glyphsLib/glyphdata.py#L216-L242

**Comment by cmyr** [comment]
> > glyphsLib is just using the glyph category/subcategory to determine the openTypeCategories?
> 
> https://github.com/googlefonts/glyphsLib/blob/e2ebf5b517d59bec0c9437da3a748c58f2999911/Lib/glyphsLib/builder/features.py#L205


Interestingly, from that code:


>  Determining the categories requires anchor propagation or user care to work
    as expected, as Glyphs.app also looks at anchors for classification


Which means that... we need to do this _after_ anchor propagation, to be strictly correct, since propagation might add an anchor and so make some glyph into a 'Base'? This also complicates doing propagation in IR..

**Comment by anthrotype** [comment]
Coming back to this two years later

>> what if UFO doesn't have either? good question. Probably I would synthesize one from GlyphData.xml? Is there a reason this is bad?

> no, that would not match fontmake and would surprise users of UFO+DS. The GlyphData.xml is very Glyphs.app specific and should only be used for .glyphs sources. The glyph names defined in there are not a standard.
If there are no categories defined, the mark writer should assume that any potential glyph that has attaching anchors is either a base, or a mark, or a ligature (if it has the _1, _2, etc. suffixed anchors).

this affects specifically MerriweatherSans.designspace, where UFOs have neither `public.openTypeCategories` nor `table GDEF` in features.fea:

```
python3 -m ttx_diff 'https://github.com/SorkinType/Merriweather-Sans?bb6bd99bf9#sources/MerriweatherSans.designspace'
```

25 glyphs (IJ, f_f, f_f_i, fi, fl, uni01C4–uni01CC, etc.) get class=2 in fontmake via this path but stay at class=1 in fontc.

The ligature class=2 assignments in fontmake's output emerge from a two-stage pipeline:

1. `ufo2ft markFeatureWriter._makeMarkToLigaAttachments()` ([markFeatureWriter.py:673-681](https://github.com/googlefonts/ufo2ft/blob/main/Lib/ufo2ft/featureWriters/markFeatureWriter.py#L673-L681)): When `gdefClasses.ligature is None` (because public.openTypeCategories is absent/empty), the filter becomes permissive:
   
   ```python
   if glyphName in markGlyphNames or (
       ligatureClass is not None and glyphName not in ligatureClass
   ):
       continue
   ```

   With ligatureClass = None, *any non-mark glyph with numbered anchors* (top_1, top_2, etc.) gets MarkToLig treatment.

2. `feaLib buildGDEFGlyphClassDef_()` ([builder.py:828-844](https://github.com/fonttools/fonttools/blob/main/Lib/fontTools/feaLib/builder.py#L828-L844)): When no explicit table GDEF block exists in the FEA, feaLib iterates compiled GPOS lookups and calls inferGlyphClasses(). MarkLigPosBuilder assigns class=2 to every glyph that appeared in a MarkToLig lookup.

So ufo2ft doesn't "know" which glyphs are ligatures but permissively emits MarkToLig for anything with numbered anchors, and GDEF class=2 falls out as a side effect of feaLib's inference from the compiled lookups.

The suggested fix for fontc would be, in the mark feature writer, when ligature class info is absent/empty, match ufo2ft's behavior: emit MarkToLig for any non-mark glyph with numbered anchors rather than requiring explicit ligature classification. This is simpler than GlyphData.xml synthesis and achieves parity with fontmake.

**Comment by anthrotype** [comment]
@cmyr Sekuya ([kevinnseptian/SEKUYA](https://github.com/kevinnseptian/SEKUYA)) is a concrete case where `prefer_gdef_categories_in_fea = false` for Glyphs sources causes a divergence. The source has a hand-authored `table GDEF` in featurePrefixes with an empty Mark slot; fontc ignores it and emits mark features, fontmake respects it and skips them.

Per my earlier comment above, since glyphsLib no longer writes `table GDEF` to FEA (since https://github.com/googlefonts/glyphsLib/pull/792 from July 2022), any such block in a Glyphs source would be considered genuinely user-authored. Should fontc still insist on setting the flag to false for Glyphs sources despite this?
You can argue that in this particular case it may a font source bug, but GIGO applies and "matches fontmake"...

**Comment by cmyr** [comment]
that's curious, is the argument for just removing the flag entirely? I could try that and do a crater run and see how it looks, if that would make sense?
